pkg/controller/direct/recaptchaenterprise/firewallpolicy_controller.go

```diff
@@ -0,0 +1,252 @@
+
+	// This resource has a server-generated ID. This means user should not know
+	// the ID before the resource is created, and 'metadata.name' won't be used
+	// as the default resource ID. So empty value for 'spec.resourceID' should
+	// also be valid:
+	// 1. When 'spec.resourceID' is not set or set to an empty value, the
+	//    intention is to create the resource.
+	// 2. When 'spec.resourceID' is set, the intention is to acquire an existing
+	//    resource.
+	//    2.1. When 'spec.resourceID' is set but the corresponding GCP resource
+	//         is not found, then it is a real error.
+	if a.id.ID() == "" {
+		log.V(2).Info("no resource ID in get indicates the create intention", "name", a.id)
+		return false, nil
+	}
+	req := &pb.GetFirewallPolicyRequest{Name: a.id.String()}
+	firewallpolicypb, err := a.gcpClient.GetFirewallPolicy(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, fmt.Errorf("firewallPolicy %q can't be acquired: %w; unset 'spec.resourceID' if you want to create it", a.id, err)
+		}
+		return false, fmt.Errorf("getting FirewallPolicy %q: %w", a.id, err)
+	}
+
+	a.actual = firewallpolicypb
+	return true, nil
+}
+
+// Create creates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *FirewallPolicyAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("creating FirewallPolicy", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := ReCAPTCHAEnterpriseFirewallPolicySpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	req := &pb.CreateFirewallPolicyRequest{
+		Parent:         a.id.Parent().String(),
+		FirewallPolicy: resource,
+	}
+	created, err := a.gcpClient.CreateFirewallPolicy(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating FirewallPolicy %s: %w", a.id, err)
+	}
+	log.V(2).Info("successfully created FirewallPolicy", "name", a.id)
+
+	status := &krm.ReCAPTCHAEnterpriseFirewallPolicyStatus{}
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	_, actualResourceID, err := krm.ParseFirewallPolicyExternal(created.Name)
+	if err != nil {
+		return fmt.Errorf("parsing the resource name in the response of CreateFirewallPolicy: %w", err)
+	}
+	status.ExternalRef = direct.LazyPtr(fmt.Sprintf("%s/firewallpolicies/%s", a.id.Parent(), actualResourceID))
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+// Update updates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *FirewallPolicyAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("updating FirewallPolicy", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	desiredPb := ReCAPTCHAEnterpriseFirewallPolicySpec_ToProto(mapCtx, &a.desired.Spec)
```

Comment:
Can you double check if user can specify `name` in the Create call. If so, this logic need to be updated to support custom id

Reply:
The resource has service-generated ID so the name shouldn't be set during Create. Even if it is set, the value won't be respected by the API.